server/api/src/rs_get_resource_info_for_operation.cpp

```diff
@@ -0,0 +1,83 @@
+#include "irods/get_resource_info_for_operation.h"
+
+#include "irods/rsGlobalExtern.hpp"
+#include "irods/rcGlobalExtern.h"
+#include "irods/getRemoteZoneResc.h"
+#include "irods/dataObjCreate.h"
+#include "irods/objMetaOpr.hpp"
+#include "irods/resource.hpp"
+#include "irods/collection.hpp"
+#include "irods/specColl.hpp"
+#include "irods/miscServerFunct.hpp"
+#include "irods/openCollection.h"
+#include "irods/readCollection.h"
+#include "irods/closeCollection.h"
+#include "irods/dataObjOpr.hpp"
+#include "irods/rsGetHostForGet.hpp"
+
+#include "irods/irods_at_scope_exit.hpp"
+#include "irods/irods_logger.hpp"
+#include "irods/irods_resource_backport.hpp"
+#include "irods/irods_resource_redirect.hpp"
+
+#include <nlohmann/json.hpp>
+#include <fmt/format.h>
+
+#include <array>
+#include <cstring>      // for strdup
+#include <string>
+
+int rs_get_resource_info_for_operation(
+    rsComm_t*     _rsComm,
+    dataObjInp_t* _dataObjInp,
+    char **       _out_info )
+{
+    using log_api = irods::experimental::log::api;
+    rodsServerHost_t *rodsServerHost = nullptr;
+    const int remoteFlag = getAndConnRemoteZone(_rsComm, _dataObjInp, &rodsServerHost, REMOTE_OPEN);
+    if (remoteFlag < 0) {
+        return remoteFlag;
+    }
+
+    if (REMOTE_HOST == remoteFlag) {
+        return rc_get_resource_info_for_operation(rodsServerHost->conn, _dataObjInp, _out_info);
+    }
+
+    std::string hier;
+    std::string location;
+    const char * op_type = getValByKey(&_dataObjInp->condInput, GET_RESOURCE_INFO_OP_TYPE_KW);
+    const std::array allowed_ops{ irods::CREATE_OPERATION,
+                                  irods::WRITE_OPERATION,
+                                  irods::UNLINK_OPERATION,
+                                  irods::OPEN_OPERATION    };
+
+    decltype(allowed_ops.end()) matched_operation;
+    if (!op_type || allowed_ops.end() == (matched_operation = std::find(allowed_ops.begin(), allowed_ops.end(), op_type))) {
+        return SYS_INVALID_INPUT_PARAM;
+    }
```

Comment:
This is close to what I had in mind, yes.

I see what you meant now about the readability of this code. To help with this, we can just separate out the `!op_type` check. Here's something closer to what I was thinking of. Let me know if this seems better or worse:
```suggestion
    const char * op_type = getValByKey(&_dataObjInp->condInput, GET_RESOURCE_INFO_OP_TYPE_KW);
    if (!op_type) {
        // Consider adding an rError message or logging something here?
        return SYS_INVALID_INPUT_PARAM;
    }
    
    const std::array allowed_ops{ irods::CREATE_OPERATION,
                                  irods::WRITE_OPERATION,
                                  irods::UNLINK_OPERATION,
                                  irods::OPEN_OPERATION    };

    auto matched_operation = std::find(allowed_ops.begin(), allowed_ops.end(), op_type);
    if (allowed_ops.end() == matched_operation) {
        // Consider adding an rError message or logging something here?
        return SYS_INVALID_INPUT_PARAM;
    }
```